### PR TITLE
Cardano lock: The verification method of the update contract

### PR DESCRIPTION
**Pull request by joii2020**
According to the documentation: https://github.com/XuJiandong/docs-bank/pull/8/files

### CODE REVIEW

File: c/cardano_lock.c [hunk 1]
```diff
@@ -225,88 +236,201 @@ int cardano_convert_copy(uint8_t *output, size_t *output_len,
...
+  return err;
+}
+
+int _get_cursor_from_witness(WitnessArgsType* witness,
+                             size_t index,
+                             size_t source) {
+  int err = 0;
+  mol2_cursor_t cur;
+  uint64_t len = 0;
+  ckb_load_witness(NULL, &len, 0, index, source);
```
Comment (XuJiandong) [inline c/cardano_lock.c:309]: check returned error code.

File: c/cardano_lock.c [hunk 2]
```diff
@@ -225,88 +236,201 @@ int cardano_convert_copy(uint8_t *output, size_t *output_len,
...
-  CHECK(err == CKB_SUCCESS, err);
-
-  uint8_t pubkey[PUBLIC_KEY_SIZE] = {0};
+  uint8_t pub_key[PUBLIC_KEY_SIZE] = {0};
   uint8_t signature[SIGNATURE_SIZE] = {0};
-  err = get_pubkey_and_signature(pubkey, signature);
+  mol2_cursor_t new_message_cursor;
+  err = get_witness_data(pub_key, signature, &new_message_cursor);
   CHECK(err == CKB_SUCCESS, err);
+  uint8_t new_message[new_message_cursor.size];
```
Comment (XuJiandong) [inline c/cardano_lock.c:412]: Check `new_message_cursor.size`: too large, e.g. ` > 65536`.

File: c/cardano_lock.c [hunk 3]
```diff
@@ -225,88 +236,201 @@ int cardano_convert_copy(uint8_t *output, size_t *output_len,
...
-  int err = generate_sighash_all(sighash_msg, sizeof(sighash_msg));
-  CHECK(err == CKB_SUCCESS, err);
+void get_pubkey_hash(uint8_t* pubkey, uint8_t* hash) {
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, pubkey, PUBLIC_KEY_SIZE);
+  blake2b_final(&blake2b_ctx, hash, BLAKE2B_BLOCK_SIZE);
+}
+
+int get_payload(const uint8_t* new_msg, size_t len, uint8_t* payload) {
```
Comment (XuJiandong) [inline c/cardano_lock.c:351]: don't forget to fuzz this function.

File: c/cardano_lock.c [hunk 4]
```diff
@@ -225,88 +236,201 @@ int cardano_convert_copy(uint8_t *output, size_t *output_len,
...
+  const char* msg_sign_context = "Signature1";
+  // msg_sign_context string size is 10
+  CHECK(tmp_len == 10, ERROR_PAYLOAD);
+  CHECK(memcmp(msg_sign_context, tmp_buf, tmp_len) == 0, ERROR_PAYLOAD);
+
+  // null
+  tmp_buf = NULL;
+  tmp_len = 0;
+  err = nanocbor_get_bstr(&n_array, (const uint8_t**)&tmp_buf, &tmp_len);
+  CHECK(err == NANOCBOR_OK, ERROR_PAYLOAD);
```
Comment (XuJiandong) [inline c/cardano_lock.c:375]: tmp_len should be zero? I think it needs to be checked.

File: tests/cardano_lock_rust/tests/test_cardano_lock.rs [hunk 5]
```diff
@@ -120,7 +73,7 @@ fn failed_sign_data() {
     verifier.set_debug_printer(debug_printer);
 
     let verify_result = verifier.verify(MAX_CYCLES);
-    assert!(verify_result.is_err());
+    assert!(verify_result.is_err(), "pass verification");
 }
 
 #[test]
```
Comment (XuJiandong) [inline tests/cardano_lock_rust/tests/test_cardano_lock.rs:79]: Add a random test cases. Fill witness lock with random data and loop 100 times: always failed.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: c/cardano_lock.c (hunk 4 above)
Comment (joii2020) [reply c/cardano_lock.c:375]: These two values are not required in this contract.

Comment (XuJiandong) [reply c/cardano_lock.c:375]: OK

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**